controllers/topology/machineset_controller.go

```diff
@@ -0,0 +1,204 @@
+	// race conditions caused by an outdated cache.
+	APIReader        client.Reader
+	WatchFilterValue string
+}
+
+func (r *MachineSetReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineSet{}).
+		Named("topology/machineset").
+		WithOptions(options).
+		WithEventFilter(predicates.All(ctrl.LoggerFrom(ctx),
+			predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue),
+			predicates.ResourceIsTopologyOwned(ctrl.LoggerFrom(ctx)),
+		)).
+		Complete(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	return nil
+}
+
+// Reconcile deletes referenced templates during deletion of topology-owned MachineSets.
+// The templates are only deleted, if they are not used in other MachineDeployments or MachineSets which are not in deleting state,
+// i.e. the templates would otherwise be orphaned after the MachineSet deletion completes.
+// Additional context:
+// * MachineSet deletion:
+//   * MachineSets are deleted and garbage collected first (without waiting until all Machines are also deleted)
+//   * After that, deletion of Machines is automatically triggered by Kubernetes based on owner references.
+// Note: We assume templates are not reused by different MachineDeployments, which is (only) true for topology-owned
+//       MachineDeployments.
+func (r *MachineSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the MachineSet instance.
+	ms := &clusterv1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, ms); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get MachineSet/%s", req.NamespacedName.Name)
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, ms.Namespace, ms.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, ms) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Set finalizer, if it's not already set.
+	if !controllerutil.ContainsFinalizer(ms, clusterv1.MachineSetTopologyFinalizer) {
+		patchHelper, err := patch.NewHelper(ms, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: ms})
+		}
+		controllerutil.AddFinalizer(ms, clusterv1.MachineSetTopologyFinalizer)
+		if err := patchHelper.Patch(ctx, ms); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: ms})
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion reconciliation loop.
```

Comment:
wouldn't this introduce a race with the garbage collector? e.g if next reconcile is fast enough it would get the finaliser back?
Should the check above be
```
if !controllerutil.ContainsFinalizer(ms, clusterv1.MachineSetTopologyFinalizer) && ms.ObjectMeta.DeletionTimestamp.IsZero() {
```
?

Reply:
In my opinion - yes. To be honest, it was on my list for follow-up discussions, but I should have raised it immediately.

We are always adding finalizers independent of the deletion timestamp (also in the regular machinedeployment and cluster controller).

I agree, we should only set the the finalizer when the resource is not in deleting state.

@fabriziopandini @vincepri WDYT?

Reply:
If we re-add a finalizer (even if it's reconciling an already-deleted resource), wouldn't the controller just do its own thing and potentially deleted it at the end?

Reply:
This is not happening in other controllers, but I don't see problem in adding the finalizer only if the object is not deleting

Reply:
We should actually keep it as-is, if a MachineDeployment is deleted fast enough we need to make sure to always have that finalizer on it, regardless if it has a deletion timestamp or not. The code should be re-entrant and reconcile the deletion and then remove the finalizer properly.

Reply:
An example how we thing the race condition can occur:

* user (or whoever) deletes the resource => deletion timestamp is set on the resource
* machinesetreconciler => reconcileDelete + remove finalizer

Now one of the following happens:
* Kubernetes garbage collector detects no finalizers are set => resource is deleted from etcd
* machinesetreconciler gets an update event for the MachineSet (because of the finalizer removal) => adds the finalizer
  * => cycle repeats until the Kubernetes garbage collector is faster

Reply:
Okay so for the current case we'll fix it by only adding the finalizer on creation. Currently working on it based on vince's latest commit (and also test it locally)

Reply:
Finalized and tested it (successfully) locally.

Reply:
Latest version is adding the finaliser on object creation so there should not be anymore risk of race condition with deletion